src/circuit.rs

```diff
@@ -116,133 +126,153 @@ impl<E: PairingEngine, P: TEModelParameters<BaseField = E::Fr>>
 /// // 3) b <= 2^5
 /// // 4) a * b = d where D is a PI
 /// // 5) JubJub::GENERATOR * e(JubJubScalar) = f where F is a PI
-/// #[derive(Debug, Default)]
-/// pub struct TestCircuit {
+/// #[derive(Debug)]
+/// pub struct TestCircuit<
+///     E: PairingEngine,
+///     P: TEModelParameters<BaseField = E::Fr>
+/// > {
 ///     a: E::Fr,
 ///     b: E::Fr,
 ///     c: E::Fr,
 ///     d: E::Fr,
-///     e: JubJubScalar,
-///     f: JubJubAffine,
+///     e: P::ScalarField,
+///     f: GroupAffine<P>,
 /// }
 ///
+///     impl<
+///         E: PairingEngine,
+///         P: TEModelParameters<BaseField = E::Fr>
+///     > Default for TestCircuit<E, P> {
+///         fn default() -> Self {
+///             Self {
+///                 a: E::Fr::zero(),
+///                 b: E::Fr::zero(),
+///                 c: E::Fr::zero(),
+///                 d: E::Fr::zero(),
+///                 e: P::ScalarField::zero(),
+///                 f: GroupAffine::<P>::zero(),
+///             }
+///         }
+///     }
+///
 /// impl<
 ///         E: PairingEngine,
-///         T: ProjectiveCurve<BaseField = E::Fr>,
 ///         P: TEModelParameters<BaseField = E::Fr>,
-///     > Circuit<E, T, P> for TestCircuit<E, T, P>
+///     > Circuit<E, P> for TestCircuit<E, P>
 /// {
 ///     const CIRCUIT_ID: [u8; 32] = [0xff; 32];
 ///     fn gadget(
 ///         &mut self,
-///         composer: &mut StandardComposer<E, T, P>,
+///         composer: &mut StandardComposer<E, P>,
 ///     ) -> Result<(), Error> {
 ///         // Add fixed witness zero
 ///         let a = composer.add_input(self.a);
 ///         let b = composer.add_input(self.b);
 ///         // Make first constraint a + b = c
-///         composer.poly_gate(
-///             a,
-///             b,
-///             composer.zero_var,
-///             E::Fr::zero(),
-///             E::Fr::one(),
-///             E::Fr::one(),
-///             E::Fr::zero(),
-///             E::Fr::zero(),
-///             Some(-self.c),
+///         composer.add(
+///           (E::Fr::zero(), a),
+///           (E::Fr::zero(), b),
+///           E::Fr::zero(),
+///           Some(-self.c),
 ///         );
 ///         // Check that a and b are in range
 ///         composer.range_gate(a, 1 << 6);
 ///         composer.range_gate(b, 1 << 5);
 ///         // Make second constraint a * b = d
-///         composer.poly_gate(
-///             a,
-///             b,
-///             composer.zero,
-///             E::Fr::one(),
-///             E::Fr::zero(),
-///             E::Fr::zero(),
-///             E::Fr::one(),
-///             E::Fr::zero(),
-///             Some(-self.d),
-///         );
+///         composer.mul(E::Fr::one(), a, b, E::Fr::zero(), Some(-self.d));
 ///
-///         let e = composer
-///             .add_input(util::from_embedded_curve_scalar::<E, P>(self.e));
+///         let e_repr = self.e.into_repr().to_bytes_le();
+///         let e = composer.add_input(E::Fr::from_le_bytes_mod_order(&e_repr));
```

Comment:
I think we can remove `e_repr` and then add

```
let e = composer.add_input(util::from_embedded_curve_scalar(self.e));

Reply:
problem is that `util` is not a public module. But I agree this does not look great for the caller

Reply:
Let's keep it how it was. And solve this on a follow-up PR refactoring a bit the `Variable` API.